### PR TITLE
Add some examples to std::string

### LINKED ISSUE #30345: No mention of concatenation in std::string::String docs
In the documentation for [std::string::String](http://doc.rust-lang.org/nightly/std/string/struct.String.html), there is no mention of concatenation - it's not obvious that the `+` operator is used for concatenation, nor that it is restricted to `String + &str`, and not, eg. `String + String`.

There should be an example showing how to concatenate/format the different string types, or, at the very least, a mention of the word 'concatenate', to allow easy searching.

### PR DESCRIPTION
**Pull request by steveklabnik**
Fixes #30345

I'm not sure if there's anything else that belongs here. Thoughts?

### CODE REVIEW

**Comment by ollie27** [comment]
I think in order to actually fix #30345 there should be a mention of the word "concatenation", perhaps in the description and examples for `push_str`.

File: src/libcollections/string.rs [hunk 1]
```diff
@@ -16,6 +16,41 @@
...
+//! let sparkle_heart = vec![240, 159, 146, 150];
+//!
+//! // We know these bytes are valid, so we'll use `unwrap()`.
+//! let sparkle_heart = String::from_utf8(sparkle_heart).unwrap();
+//!
+//! assert_eq!("💖", sparkle_heart);
+//!
+//! let bytes = sparkle_heart.into_bytes();
+//!
+//! assert_eq([240, 159, 146, 150, bytes);
```
Comment (ollie27) [inline src/libcollections/string.rs]: Should probably be:
```rust
assert_eq!(bytes, [240, 159, 146, 150]);
```

Comment (steveklabnik) [inline src/libcollections/string.rs]: For what reason?

**Comment by steveklabnik** [comment]
Good call on putting the word 'concatenation'

File: src/libcollections/string.rs [hunk 2]
```diff
@@ -16,6 +16,42 @@
...
+//! let sparkle_heart = vec![240, 159, 146, 150];
+//!
+//! // We know these bytes are valid, so we'll use `unwrap()`.
+//! let sparkle_heart = String::from_utf8(sparkle_heart).unwrap();
+//!
+//! assert_eq!("💖", sparkle_heart);
+//!
+//! let bytes = sparkle_heart.into_bytes();
+//!
+//! assert_eq([240, 159, 146, 150], bytes);
```
Comment (durka) [inline src/libcollections/string.rs]: !

Comment (ollie27) [inline src/libcollections/string.rs]: and the arguments still need to be swapped because PartialEq<Vec<u8>> is not implemented for [u8; 4]

Comment (steveklabnik) [inline src/libcollections/string.rs]: ugh, how did this pass for me locally, then?

Comment (durka) [inline src/libcollections/string.rs]: What's your test setup / make command?

**Comment by steveklabnik** [comment]
@bors: r=brson rollup

:(

**Pull request merged**